internal/controller/etcdopstask/reconciler.go

```diff
@@ -0,0 +1,153 @@
+func (r *Reconciler) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+	logger := r.logger.WithValues(
+		"runId", string(controller.ReconcileIDFromContext(ctx)),
+		"namespace", req.Namespace,
+		"name", req.Name,
+	)
+	logger.Info("Reconciling EtcdOperatorTask")
+
+	task, err := r.getTask(ctx, req.NamespacedName)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			logger.Info("Task not found, skipping reconciliation",
+				"namespace", req.Namespace,
+				"name", req.Name)
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+	if task == nil {
+		// NotFound: resource deleted or not created yet, do not requeue
+		logger.Info("Task not found, skipping reconciliation",
+			"namespace", req.Namespace,
+			"name", req.Name)
+		return reconcile.Result{}, nil
+	}
+
+	taskHandlerInstance, err := r.createTaskHandlerInstance(task, logger)
+	if err != nil {
+		logger.Error(err, "Failed to create task handler instance")
+		return reconcile.Result{}, nil
+	}
+	logger.Info(fmt.Sprintf("Task handler created successfully: %T", taskHandlerInstance))
+
+	if r.shouldDeleteTask(task) {
+		return r.triggerDeletionFlow(ctx, taskHandlerInstance, task).ReconcileResult()
+	}
+
+	result := r.reconcileTask(ctx, client.ObjectKeyFromObject(task), taskHandlerInstance)
+	if result.HasErrors() || result.NeedsRequeue() {
+		return result.ReconcileResult()
+	}
+
+	task, err = r.getTask(ctx, req.NamespacedName)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	if task.IsCompleted() {
+		return reconcile.Result{RequeueAfter: task.GetTimeToExpiry()}, nil
+	}
+	return reconcile.Result{}, nil
+}
+
+// createTaskHandlerInstance instantiates the appropriate TaskHandler for the given task.
+func (r *Reconciler) createTaskHandlerInstance(task *v1alpha1.EtcdOpsTask, logger logr.Logger) (task.Handler, error) {
+	configValue := reflect.ValueOf(task.Spec.Config)
+	configType := reflect.TypeOf(task.Spec.Config)
+	taskType := ""
+
+	for i := 0; i < configValue.NumField(); i++ {
+		field := configValue.Field(i)
+		fieldType := configType.Field(i)
+
+		if field.Kind() == reflect.Ptr && !field.IsNil() {
+			taskType = fieldType.Name
+			break
+		}
+	}
+	return r.taskHandlerRegistry.GetHandler(taskType, r.client, logger, task, nil)
+}
+
+func (r *Reconciler) shouldDeleteTask(task *v1alpha1.EtcdOpsTask) bool {
+	return task.IsCompleted() || task.IsMarkedForDeletion()
+}
+
+// createAndInitializeTaskHandlerRegistry creates and initializes the task handler registry with default handlers.
+func createAndInitializeTaskHandlerRegistry() task.TaskHandlerRegistry {
+	registry := task.NewTaskHandlerRegistry()
+
+	// Register OnDemandSnapshot handler
+	registry.Register("OnDemandSnapshot", func(k8sclient client.Client, logger logr.Logger, task *v1alpha1.EtcdOpsTask, httpClient *http.Client) (task.Handler, error) {
+		return ondemandsnapshot.NewWithHTTPClient(k8sclient, logger, task, httpClient)
+	})
```

Comment:
The in-line function wasn't necessary, it has the same signature as `ondemandsnapshot.NewWithHTTPClient`. The function value can be passed directly making it much simpler.

```suggestion
	registry.Register("OnDemandSnapshot", ondemandsnapshot.NewWithHTTPClient)
```